R/formula.R

```diff
@@ -0,0 +1,422 @@
+  if (!global_intercept) {
+    allterms_nobias <- unlist(lapply(allterms, function(term) paste0(term, " + 0")))
+    return(allterms_nobias)
+  } else
+    return(allterms)
+}
+
+split_interaction_term <- function(term) {
+  ## strong heredity by default
+  variables <- unlist(strsplit(term, ":"))
+  individual_joint <- paste(variables, collapse = " + ")
+  joint_intercept <- paste(c(individual_joint, term), collapse = " + ")
+
+  joint_intercept
+}
+
+split_group_term <- function(term) {
+  ## this expands whatever terms() did not expand
+  term <- gsub("\\)$", "",
+               gsub("^\\(", "",
+                    flatten_group_terms(term)))
+  if ("\\-" %in% term)
+    stop(paste0("Use of `-` is not supported, omit variables or use the ",
+                "method update on the formula, or write `0 +` to remove ",
+                "the intercept."))
+
+  chunks <- strsplit(term, " \\| ")[[1]]
+  variables <- strsplit(chunks[1], " \\+ ")[[1]]
+
+  ## split possible interaction terms
+  int_t <- str_match(variables, ":")
+  ## int_v <- lapply(variables[!is.na(int_t)], split_interaction_term)
+  int_v <- variables[!is.na(int_t)]
+  lin_v <- variables[is.na(int_t)]
+
+  ## variables <- c(int_v, lin_v)
+
+  ## don't add the intercept twice
+  variables <- setdiff(variables, "1")
+  group <- chunks[2]
+
+  if ("0" %in% variables)
+    group_intercept <- FALSE
+  else
+    group_intercept <- TRUE
+
+  group_terms <- lapply(lin_v, function(v) paste0(v, " + ", "(0 + ", v, " | ", group, ")"))
+  group_terms <- c(group_terms, lapply(int_v, function(v)
+    paste0(split_interaction_term(v), " + ", "(0 + ", v, " | ", group, ")")))
+
+  if (group_intercept) {
+    group_terms <- c(group_terms, list(paste0("(1 | ", group, ")")))
+    group_terms <- c(group_terms,
+                     lapply(lin_v,
+                            function(v)
+                              paste0(v, " + ", "(", v, " | ", group, ")")))
+    group_terms <- c(group_terms,
+                     lapply(int_v,
+                            function(v)
+                              paste0(split_interaction_term(v), " + ", "(", v, " | ", group, ")")))
+
+    ## add v + ( 1 | group)
+    group_terms <- c(group_terms,
+                     lapply(lin_v,
+                            function(v)
+                              paste0(v, " + ", "(1 | ", group, ")")))
+    group_terms <- c(group_terms,
+                     lapply(int_v,
+                            function(v)
+                              paste0(split_interaction_term(v), " + ", "(1 | ", group, ")")))
+  }
+
+  group_terms
+}
+
+formula_contains_group_terms <- function(formula) {
+  t <- terms(formula)
+  attributes <- attributes(t)
+  terms <- attributes$term.labels
+
+  hier <- str_match(terms, "\\|")
+  group_terms <- terms[!is.na(hier)]
+
+  return(length(group_terms) > 0)
+}
+
+subset_formula <- function(formula, terms, data, y=NULL, return_list=FALSE) {
```

Comment:
Okay, it can be more verbose. I'll rename the argument to `split_formulas`. The intention of the argument is to return a list of formulas where each of them has a single response, instead of a formula with multiple responses.